### PR TITLE
Switched to AsyncKeyedLock

### PR DESCRIPTION
**Pull request by MarkCiliaVincenti**
Fixed problems with InMemoryDistributedLock that was allowing concurrent threads for keys and not clearing from a dictionary.

### CODE REVIEW

**Comment by MarkCiliaVincenti** [comment]
@Coldairarrow

**Comment by MarkCiliaVincenti** [comment]
@Coldairarrow ?

File: tests/Colder.Tests/InMemoryDistributedLockTest.cs [hunk 1]
```diff
@@ -36,10 +37,10 @@ public async Task Test()
             {
                 tasks.Add(Task.Factory.StartNew(async () =>
                 {
-                    for (int j = 0; j < 10000; j++)
+                    for (int j = 0; j < 100; j++)
                     {
                         using var _ = await theLock.Lock(key);
-                        num++;
+                        Interlocked.Increment(ref num);
```
Comment (Coldairarrow) [inline tests/Colder.Tests/InMemoryDistributedLockTest.cs:43]: Interlocked is thread safe,but this code is to test the distributedlock is work fine or not

File: tests/Colder.Tests/InMemoryDistributedLockTest.cs [hunk 2]
```diff
@@ -49,7 +50,7 @@ public async Task Test()
                 await (await aTask);
             }
 
-            Assert.AreEqual(num, 160000);
+            Assert.AreEqual(1600, num);
```
Comment (Coldairarrow) [inline tests/Colder.Tests/InMemoryDistributedLockTest.cs:53]: Interlocked is thread safe,but this code is to test the distributedlock is work fine or not,so i think there is unnecessary

**Comment by MarkCiliaVincenti** [comment]
OK @Coldairarrow ?

**Comment by MarkCiliaVincenti** [comment]
@Coldairarrow ?

**Comment by MarkCiliaVincenti** [comment]
@Coldairarrow ?

**Comment by Coldairarrow** [comment]
```c#
using AsyncKeyedLock;

namespace Test
{
    class Program
    {
        public static async Task Main(string[] args)
        {
            AsyncKeyedLocker<string> _asyncKeyedLocker = new(o =>
            {
                o.PoolSize = 20;
                o.PoolInitialFill = 1;
            });

            var theLock = _asyncKeyedLocker;
            string key = Guid.NewGuid().ToString();

            List<Task<Task>> tasks = new List<Task<Task>>();

            int num = 0;
            for (int i = 0; i < 16; i++)
            {
                tasks.Add(Task.Factory.StartNew(async () =>
                {
                    for (int j = 0; j < 1000; j++)
                    {
                        using var _ = await theLock.LockAsync(key);
                        num++;
                    }
                }, TaskCreationOptions.LongRunning));
            }

            await Task.WhenAll(await Task.WhenAll(tasks));

            Console.WriteLine(num == 16000);
        }
    }
}
```
this code is not work fine,the cpu is 100%,so is there any problem?

**Comment by Coldairarrow** [comment]
@MarkCiliaVincenti

**Comment by MarkCiliaVincenti** [comment]
> ```cs
> using AsyncKeyedLock;
> 
> namespace Test
> {
>     class Program
>     {
>         public static async Task Main(string[] args)
>         {
>             AsyncKeyedLocker<string> _asyncKeyedLocker = new(o =>
>             {
>                 o.PoolSize = 20;
>                 o.PoolInitialFill = 1;
>             });
> 
>             var theLock = _asyncKeyedLocker;
>             string key = Guid.NewGuid().ToString();
> 
>             List<Task<Task>> tasks = new List<Task<Task>>();
> 
>             int num = 0;
>             for (int i = 0; i < 16; i++)
>             {
>                 tasks.Add(Task.Factory.StartNew(async () =>
>                 {
>                     for (int j = 0; j < 1000; j++)
>                     {
>                         using var _ = await theLock.LockAsync(key);
>                         num++;
>                     }
>                 }, TaskCreationOptions.LongRunning));
>             }
> 
>             await Task.WhenAll(await Task.WhenAll(tasks));
> 
>             Console.WriteLine(num == 16000);
>         }
>     }
> }
> ```
> 
> this code is not work fine,the cpu is 100%,so is there any problem?

I think threadpool starvation. If you change that 16 to 2 and that 1000 to 8000 it will work.
You may want to decide to use striped async locking instead to avoid this issue you are pointing.

````
        [TestMethod]
        public async Task TestAsyncKeyedLock()
        {
            StripedAsyncKeyedLocker<string> _asyncKeyedLocker = new();

            var theLock = _asyncKeyedLocker;
            string key = Guid.NewGuid().ToString();

            List<Task<Task>> tasks = new List<Task<Task>>();

            int num = 0;
            for (int i = 0; i < 16; i++)
            {
                tasks.Add(Task.Factory.StartNew(async () =>
                {
                    for (int j = 0; j < 10000; j++)
                    {
                        using (await theLock.LockAsync(key).ConfigureAwait(false))
                        {
                            ++num;
                        }
                    }
                }, TaskCreationOptions.LongRunning));
            }

            await Task.WhenAll(await Task.WhenAll(tasks));
        }
````

Read here for the comparison between the current locking and the striped locking: https://github.com/MarkCiliaVincenti/AsyncKeyedLock/wiki

**Comment by MarkCiliaVincenti** [comment]
Also if you simplify your code, it works well.

Replace:
````
            List<Task<Task>> tasks = new List<Task<Task>>();

            int num = 0;
            for (int i = 0; i < 16; i++)
            {
                tasks.Add(Task.Factory.StartNew(async () =>
                {
                    for (int j = 0; j < 1000; j++)
                    {
                        using var _ = await theLock.LockAsync(key);
                        num++;
                    }
                }, TaskCreationOptions.LongRunning));
            }

            await Task.WhenAll(await Task.WhenAll(tasks));
````

with:

````
            int num = 0;

            var tasks = Enumerable.Range(1, 16000)
                .Select(async i =>
                {
                    using (await theLock.LockAsync(key))
                    {
                        ++num;
                    }
                });
            await Task.WhenAll(tasks.AsParallel());
````

**Comment by Coldairarrow** [comment]
ok, this is good

**Comment by MarkCiliaVincenti** [comment]
> ok, this is good

Released version 6.2.1 in the meantime :)

**Pull request merged**